core/src/main/java/org/apache/iceberg/LazyImmutableMap.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import org.apache.iceberg.relocated.com.google.common.collect.Maps;
+
+public class LazyImmutableMap<K, V> implements Map<K, V>, Serializable {
+
+  private final Map<K, V> copiedMap;
+  private Map<K, V> immutableMap;
+
+  LazyImmutableMap() {
+    this.copiedMap = Maps.newLinkedHashMap();
+  }
+
+  private LazyImmutableMap(Map<K, V> map) {
+    this.copiedMap = Maps.newLinkedHashMap();
```

Comment:
Why use `LinkedHashMap`? Do we need to preserve key order for some reason?

Reply:
OK, I was thinking that the map assertion will check the pairs insert order, so I used the `LinkedHashMap` here.   Read the code again, here we can use the `HashMap` directly.